File: src/platform/tizen/waitForHlsManifest.ts
/**
 * Poll until an HLS playlist is playable: #EXTM3U present and (optionally)
 * the first media segment is fetchable. AVPlay open on a not-yet-written
 * remux/transcode pipeline fails the whole player; encoded sessions also
 * expose a synthetic VOD playlist before the window-head segment exists, so
 * checking only #EXTM3U is not enough.
 *
 * Throws TranscodeStartupTimeoutError when the deadline elapses without a
 * ready segment so callers can show "Transcode timed out" instead of hanging.
 */

export class TranscodeStartupTimeoutError extends Error {
  constructor(message = "Transcode timed out") {
    super(message);
    this.name = "TranscodeStartupTimeoutError";
  }
}

export interface WaitForHlsManifestOptions {
  intervalMs?: number;
  timeoutMs?: number;
  fetchImpl?: typeof fetch;
  /** When true (default), require the first media segment to return HTTP 200. */
  requireSegment?: boolean;
  /** Fired periodically during the wait (e.g. POST /playback/.../progress). */
  onKeepAlive?: () => void | Promise<void>;
  keepAliveEveryMs?: number;
  /**
   * Stops polling (and its keepalives) when aborted. Without it, navigating away
   * from a slow-starting transcode leaves this loop running and posting progress
   * keepalives that actively hold the server session alive until it times out.
   */
  signal?: AbortSignal;
}

async function fetchText(
  url: string,
  fetchImpl: typeof fetch,
  abortMs: number,
): Promise<{ ok: boolean; status: number; body: string }> {
  try {
    const controller = new AbortController();
    const timer = window.setTimeout(() => controller.abort(), abortMs);
    try {
      const res = await fetchImpl(url, {
        cache: "no-store",
        signal: controller.signal,
      });
      const body = res.ok ? await res.text() : "";
      return { ok: res.ok, status: res.status, body };
    } finally {
      window.clearTimeout(timer);
    }
  } catch {
    return { ok: false, status: 0, body: "" };
  }
}

async function segmentReady(url: string, fetchImpl: typeof fetch): Promise<boolean> {
  try {
    const controller = new AbortController();
    const timer = window.setTimeout(() => controller.abort(), 12_000);
    try {
      // Prefer a cheap probe; some stacks reject HEAD — fall back to GET.
      let res = await fetchImpl(url, {
        method: "HEAD",
        cache: "no-store",
        signal: controller.signal,
      });
      if (res.status === 405 || res.status === 501) {
        res = await fetchImpl(url, {
          method: "GET",
          cache: "no-store",
          signal: controller.signal,
          headers: { Range: "bytes=0-0" },
        });
      }
      return res.ok || res.status === 206;
    } finally {
      window.clearTimeout(timer);
    }
  } catch {
    return false;
  }
}

/** Resolve the first media segment URI from an m3u8 body relative to the playlist URL. */
export function firstMediaSegmentUrl(playlistUrl: string, body: string): string | null {
  for (const rawLine of body.split("\n")) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#")) continue;
    try {
      const resolved = new URL(line, playlistUrl);
      // Relative segment lines omit the playlist query (auth token). Carry it
      // over so segment probes use the same credentials as the manifest.
      if (!resolved.search) {
        const base = new URL(playlistUrl);
        if (base.search) resolved.search = base.search;
      }
      return resolved.toString();
    } catch {
      return null;
    }
  }
  return null;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => window.setTimeout(resolve, ms));
}

/**
 * Wait until the HLS playlist (and optionally first segment) is ready.
 * Resolves true when ready; throws TranscodeStartupTimeoutError on timeout
 * when `throwOnTimeout` is true (default). When throwOnTimeout is false,
 * resolves false on timeout (legacy AVPlay soft-continue behavior).
 */
export async function waitForHlsManifest(
  url: string,
  options: WaitForHlsManifestOptions & { throwOnTimeout?: boolean } = {},
): Promise<boolean> {
  const intervalMs = options.intervalMs ?? 500;
  const timeoutMs = options.timeoutMs ?? 90_000;
  const fetchImpl = options.fetchImpl ?? fetch;
  const requireSegment = options.requireSegment ?? true;
  const keepAliveEveryMs = options.keepAliveEveryMs ?? 10_000;
  const throwOnTimeout = options.throwOnTimeout ?? false;
  const deadline = Date.now() + timeoutMs;
  let delay = intervalMs;
  let nextKeepAliveAt = Date.now();

  while (Date.now() < deadline) {
    // Bail the moment the caller navigates away so we stop holding the session.
    if (options.signal?.aborted) return false;
    if (options.onKeepAlive && Date.now() >= nextKeepAliveAt) {
      try {
        await options.onKeepAlive();
      } catch {
        // Keepalive is best-effort — never abort readiness polling on it.
      }
      nextKeepAliveAt = Date.now() + keepAliveEveryMs;
    }

    const { ok, body } = await fetchText(url, fetchImpl, 8_000);
    if (ok && body.includes("#EXTM3U")) {
      if (!requireSegment) return true;
      // Encoded sessions serve a synthetic VOD playlist immediately; wait until
      // the first listed segment exists so AVPlay does not hang on open.
      if (!body.includes("#EXTINF")) {
        // Empty / master-only playlist — keep polling.
      } else {
        const segmentUrl = firstMediaSegmentUrl(url, body);
        if (segmentUrl && (await segmentReady(segmentUrl, fetchImpl))) {
          return true;
        }
      }
    }

    await sleep(delay);
    delay = Math.min(Math.round(delay * 1.5), 4_000);
  }

  if (throwOnTimeout) {
    throw new TranscodeStartupTimeoutError();
  }
  return false;
}

export function isHlsUrl(url: string): boolean {
  const path = url.split("?")[0]?.toLowerCase() ?? "";
  return path.endsWith(".m3u8") || path.includes("/hls") || path.includes("master.m3u8");
}

File: flutter/packages/prairie_core/lib/src/api/wait_for_hls_manifest.dart
import 'dart:async';
import 'dart:developer' as developer;

import 'package:dio/dio.dart';

/// Thrown when the HLS playlist never becomes playable before the deadline.
/// Mirrors `TranscodeStartupTimeoutError`.
class TranscodeStartupTimeoutError implements Exception {
  TranscodeStartupTimeoutError([this.message = 'Transcode timed out']);
  final String message;

  @override
  String toString() => message;
}

/// Thrown when the readiness probe gets a 401/403 back. Unlike a 404/5xx,
/// this can never resolve by waiting longer, so it must not be swallowed
/// into the same "not ready yet" bucket that drives the poll/backoff loop.
class HlsProbeAuthError implements Exception {
  HlsProbeAuthError(this.statusCode, [this.message = 'Not authorized to fetch HLS playlist']);
  final int statusCode;
  final String message;

  @override
  String toString() => '$message (HTTP $statusCode)';
}

/// Resolve the first media segment URI from an m3u8 body relative to [playlistUrl].
String? firstMediaSegmentUrl(String playlistUrl, String body) {
  for (final rawLine in body.split('\n')) {
    final line = rawLine.trim();
    if (line.isEmpty || line.startsWith('#')) continue;
    try {
      final base = Uri.parse(playlistUrl);
      final resolved = base.resolve(line);
      if (resolved.query.isEmpty && base.query.isNotEmpty) {
        return resolved.replace(query: base.query).toString();
      }
      return resolved.toString();
    } catch (_) {
      return null;
    }
  }
  return null;
}

bool isHlsUrl(String url) {
  final path = url.split('?').first.toLowerCase();
  return path.endsWith('.m3u8') || path.contains('/hls') || path.contains('master.m3u8');
}

/// Poll until an HLS playlist is playable: `#EXTM3U` present and (optionally)
/// the first media segment is fetchable. Opening AVPlay/PlusPlayer on a
/// not-yet-written remux/transcode pipeline fails with
/// "streaming connection timed out".
Future<bool> waitForHlsManifest(
  String url, {
  Dio? dio,
  Duration interval = const Duration(milliseconds: 500),
  Duration timeout = const Duration(seconds: 90),
  bool requireSegment = true,
  bool throwOnTimeout = true,
  Duration keepAliveEvery = const Duration(seconds: 10),
  FutureOr<void> Function()? onKeepAlive,
  CancelToken? cancelToken,
}) async {
  final client = dio ?? Dio();
  final deadline = DateTime.now().add(timeout);
  var delay = interval;
  var nextKeepAliveAt = DateTime.now();

  while (DateTime.now().isBefore(deadline)) {
    if (cancelToken?.isCancelled == true) return false;

    if (onKeepAlive != null && !DateTime.now().isBefore(nextKeepAliveAt)) {
      try {
        await onKeepAlive();
      } catch (_) {
        // Keepalive is best-effort.
      }
      if (cancelToken?.isCancelled == true) return false;
      nextKeepAliveAt = DateTime.now().add(keepAliveEvery);
    }

    final body = await _fetchText(client, url, cancelToken);
    if (cancelToken?.isCancelled == true) return false;

    if (body != null && body.contains('#EXTM3U')) {
      if (!requireSegment) return true;
      if (body.contains('#EXTINF')) {
        final segmentUrl = firstMediaSegmentUrl(url, body);
        if (segmentUrl != null) {
          final ready = await _segmentReady(client, segmentUrl, cancelToken);
          if (cancelToken?.isCancelled == true) return false;
          if (ready) return true;
        }
      }
    }

    await Future<void>.delayed(delay);
    if (cancelToken?.isCancelled == true) return false;
    final nextMs = (delay.inMilliseconds * 1.5).round().clamp(interval.inMilliseconds, 4000);
    delay = Duration(milliseconds: nextMs);
  }

  if (throwOnTimeout) throw TranscodeStartupTimeoutError();
  return false;
}

/// Fetches [url] for the readiness poll. Returns the body when ready, `null`
/// when the playlist genuinely isn't written yet (404/5xx/transport error —
/// worth retrying), and throws [HlsProbeAuthError] on 401/403, which is a
/// permanent rejection no amount of polling will resolve.
Future<String?> _fetchText(Dio dio, String url, CancelToken? cancelToken) async {
  Response<String>? response;
  try {
    response = await dio.get<String>(
      url,
      cancelToken: cancelToken,
      options: Options(
        responseType: ResponseType.plain,
        receiveTimeout: const Duration(seconds: 8),
        sendTimeout: const Duration(seconds: 8),
        validateStatus: (_) => true,
      ),
    );
  } catch (_) {
    return null;
  }

  final status = response.statusCode;
  if (status == 401 || status == 403) throw HlsProbeAuthError(status!);
  if (status == null || status >= 400) {
    if (status == 404) {
      developer.log('HLS playlist not found yet (404): $url', name: 'prairie.hls_probe');
    }
    return null;
  }
  return response.data;
}

Future<bool> _segmentReady(Dio dio, String url, CancelToken? cancelToken) async {
  try {
    var response = await dio.head<void>(
      url,
      cancelToken: cancelToken,
      options: Options(
        receiveTimeout: const Duration(seconds: 12),
        sendTimeout: const Duration(seconds: 12),
        validateStatus: (_) => true,
      ),
    );
    if (response.statusCode == 405 || response.statusCode == 501) {
      response = await dio.get<void>(
        url,
        cancelToken: cancelToken,
        options: Options(
          headers: {'Range': 'bytes=0-0'},
          receiveTimeout: const Duration(seconds: 12),
          sendTimeout: const Duration(seconds: 12),
          validateStatus: (_) => true,
        ),
      );
    }
    final status = response.statusCode ?? 0;
    return status >= 200 && status < 300;
  } catch (_) {
    return false;
  }
}
